Introduce `UnvalidatedEnumValueOfInvocation` check

ghokun opened the pull request:
## Context
`Enum.valueOf` method usage may lead to runtime exceptions when the provided argument is not checked against all possible values of method owner.

Example:
```java
class Test {
  enum A {
    ONE, TWO, THREE
  }
  
  enum B {
    ONE, TWO, THREE, FOUR
  }
  
  // Will throw a runtime exception if b is equal to FOUR.
  A convertUnsafe(B b) {
    return A.valueOf(b.name());
  }
}
```

## Scope
Of course it is not able to capture all possible checks on `b` before supplying it into `A.valueOf`. Therefore scope will be limited to following checks.

```java
// BUG: Raw string
A convertUnsafe(String raw) {
  return A.valueOf(raw);
}

// BUG: From invalid string constant
A convertUnsafe() {
  return A.valueOf("FOUR");
}

// BUG: Where B has values that A do not have
A convertUnsafe(B b) {
  return A.valueOf(b.name());
}

// BUG: Value from case is not present in A
A convertUnsafe(B b) {
  return switch(b) { case FOUR -> A.valueOf(b.name()); };
}

// NOT BUG: Value from case is present in A
A convertUnsafe(B b) {
  return switch(b) { case ONE -> A.valueOf(b.name()); };
}

// NOT BUG: From valid string constant
A convertSafe() {
  return A.valueOf("ONE");
}
```

Above checks should also consider the polymorphic variant of `valueOf`:
```
public static <T extends Enum<T>> T valueOf(Class<T> enumClass, String name)
```

ghokun committed:
Add bug pattern for unchecked enum creation with valueOf

rickie committed:
Apply suggestions from EP

rickie committed:
Some late nate random refactoring

ghokun committed:
Rename class, extend checks

ghokun committed:
Add possible valid scenarios

ghokun committed:
Apply error prone suggestions

ghokun committed:
Add valueOf polymorphic variant

ghokun committed:
Check if invoked enum is filtered by a switch statement

ghokun committed:
Add check if b of b.name is used in switch (b)

ghokun committed:
Improve javadoc

ghokun committed:
Remove redundant cases

ghokun committed:
Kill more mutants

ghokun committed:
Replace type toString call

ghokun committed:
Format test code

ghokun committed:
Remove warning

ghokun committed:
Use interface where possible

ghokun committed:
more mutants

ghokun committed:
more mutants

ghokun committed:
Add default switch case handling

ghokun committed:
Fix error prone warnings

ghokun committed:
Add descriptions

ghokun committed:
Replace contains with matches

ghokun commented:
@rickie This one is ready for review now.

ghokun commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/EnumValueOfUsage.java:
```diff
@@ -0,0 +1,164 @@
+    // Match constants
+    String constantValue = ASTHelpers.constValue(nameArgument, String.class);
+    if (constantValue != null && !valuesOfSource.contains(constantValue)) {
+      return buildDescription(tree)
```
I don't think we can introduce a suggested fix for this. Description already highlights possible values.

ghokun committed:
Add more test cases

ghokun committed:
Fix formatting

rickie commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,165 @@
+    ExpressionTree nameArgument = extractNameArgument(tree, state);
+    ImmutableSet<String> valuesOfSource = findEnumValuesOfReceiver(enumType);
+
+    // Match constants
```
We generally don't need comments :)

rickie commented on error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocationTest.java:
```diff
@@ -0,0 +1,117 @@
+            "    // BUG: Diagnostic matches: AVOID_RAW",
+            "    A.valueOf(raw);",
+            "    // BUG: Diagnostic matches: AVOID_RAW",
+            "    A.valueOf(A.class, raw);",
```
I'm slightly re-ordering some of the test methods in this class

rickie committed:
Suggestions

rickie reviewed (commented):
Thanks for the cool check! I did a first pass and added a commit with some suggestions 😄.

I didn't put every change I made on as comment on the PR as I was mostly reviewing while I was flying ✈️ and didn't have internet.
If you have any questions about why I made a certain suggestion, let me know :).

The PR is already shaping up nicely!

rickie committed:
Tweak

ghokun committed:
Cover  name argument

ghokun commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,181 @@
+    if (optionalNameArgument.isEmpty()) {
+      return buildDescription(tree)
+          .setMessage(
+              "No `String` typed `name` argument was found on %s"
```
I am open to suggestions on the message for this one.

ghokun committed:
Kill null mutants

ghokun committed:
Extract name argument by position

ghokun commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    Type enumType = ASTHelpers.getReceiverType(tree);
+    ExpressionTree nameArgument = tree.getArguments().getLast();
```
@rickie Let me know what do you think about this approach.

```
A.valueOf(name)
A.valueOf(A.class, name)
```

`name` is basically the last argument in all possible cases. This allows capturing `null` easily, as we do not filter by string type.

ghokun replied:
One thing that I have not verified is, if `tree.getArguments()` is ordered or not. So far it seems so.

Stephan202 replied:
Indeed `tree.getArguments()` is ordered. However, my [comment](https://github.com/PicnicSupermarket/error-prone-support/pull/2031/changes#r3201679055) above shows that we need to handle the instance method `Enum#valueOf(String)` and `Enum#valueOf(Class, String)` distinctly anyway. Having two `Matcher`s for those automatically causes us to skip "fake" `valueOf` implementations.

ghokun replied:
Made matchers explicit.

rickie committed:
Suggestions

rickie committed:
Tweak

rickie committed:
Replace `unchecked` with `unvalidated` in `@BugPattern` summary

The term "unchecked" is overloaded in Java (unchecked casts, unchecked
exceptions). Use "unvalidated" to more precisely describe the intent.

rickie committed:
Small improvements

ghokun commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = FRAGILE_CODE)
+public final class UncheckedEnumValueOfInvocation extends BugChecker
```
```suggestion
public final class UnvalidatedEnumValueOfInvocation extends BugChecker
```

ghokun commented on error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocationTest.java:
```diff
@@ -0,0 +1,136 @@
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.jupiter.api.Test;
+
+final class UncheckedEnumValueOfInvocationTest {
```
```suggestion
final class UnvalidatedEnumValueOfInvocationTest {
```

ghokun commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Avoid passing unvalidated arguments to `Enum#valueOf`",
+    link = BUG_PATTERNS_BASE_URL + "UncheckedEnumValueOfInvocation",
```
```suggestion
    link = BUG_PATTERNS_BASE_URL + "UnvalidatedEnumValueOfInvocation",
```

ghokun commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UncheckedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+import tech.picnic.errorprone.utils.SourceCode;
+
+/**
+ * A {@link BugChecker} that flags {@link Enum#valueOf} invocations that contain unchecked
```
```suggestion
 * A {@link BugChecker} that flags {@link Enum#valueOf} invocations that contain unvalidated
```

ghokun committed:
Rename `Unchecked` to `Unvalidated`

ghokun committed:
Rename all references to

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+      return Description.NO_MATCH;
+    }
+
+    Type enumType = ASTHelpers.getReceiverType(tree);
```
The comment below discusses `A.valueOf(A.class, name)`, but as that's a static method, it should be invoked as `Enum.valueOf(A.class, name)`. In that case this statement does not correctly determine the enum type.

ghokun replied:
Added explicit matchers for all cases:

```
A.valueOf(name);
A.valueOf(A.class, name);
Enum.valueOf(A.class, name);
```

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+
+    Type enumType = ASTHelpers.getReceiverType(tree);
+    ExpressionTree nameArgument = tree.getArguments().getLast();
+    if (!MoreASTHelpers.isStringTyped(nameArgument, state)) {
+      return buildDescription(tree)
+          .setMessage(
+              "`%s` is not a valid type for `%s`, expected: `String`"
+                  .formatted(ASTHelpers.getType(nameArgument), enumType))
+          .build();
+    }
```
If we introduce the two `Matcher`s mentioned above, then we can skip this check.

ghokun replied:
This still captures `null`s. Such as:


```
"    // BUG: Diagnostic matches: INVALID_TYPE",
"    A.valueOf(null);",
"    // BUG: Diagnostic matches: INVALID_TYPE",
"    A.valueOf(A.class, null);",
```

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+  private static final Matcher<ExpressionTree> ENUM_VALUE_OF_METHOD =
+      staticMethod().onDescendantOf(Enum.class.getCanonicalName()).named("valueOf");
+  private static final Matcher<ExpressionTree> ENUM_NAME_OR_TO_STRING_METHOD =
+      instanceMethod().onDescendantOf(Enum.class.getCanonicalName()).namedAnyOf("name", "toString");
```
We should test that these methods don't take any argument.

ghokun replied:
Added `.withNoParameters();` to end.

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    // Match identifiers.
+    return nameArgument instanceof IdentifierTree ? describeMatch(tree) : Description.NO_MATCH;
```
How confident are we that this doesn't yield significantly more false positives than that it flags bugs? Even in cases where a bug is found, the path towards resolving the warning involves more than fixing the domain of values taken on by the identified variable; avoiding suppression also requires rewriting the code in one of the shapes covered further up. As-is, users are not informed of that. TBH, I suspect we'll just not want to flag this. 🤔

ghokun replied:
These are currently matching

```
"    // BUG: Diagnostic contains:",
"    A.valueOf(raw);",
"    // BUG: Diagnostic contains:",
"    A.valueOf(A.class, raw);",
"    // BUG: Diagnostic contains:",
"    Enum.valueOf(A.class, raw);",
```

We can leave it up to the consumers that they have validated this raw string outside manually and remove this check completely.

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+  private static ImmutableSet<String> findSwitchCoveredValues(
+      ExpressionTree enumValueArgument, ImmutableSet<String> valuesOfReceiver, VisitorState state) {
+    CaseTree enclosingCaseTree = ASTHelpers.findEnclosingNode(state.getPath(), CaseTree.class);
+    /* Fast path: avoids unnecessary switch-tree lookups when not inside a case. */
```
Usually we place this "fast path" comment inside the branch.

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+      switchCases = switchStmtTree.getCases();
+    }
+
+    Type nameInvocationReceiverType = ASTHelpers.getReceiverType(enumValueArgument);
```
This logic assumes that `enumValueArgument` matches `ENUM_NAME_OR_TO_STRING_METHOD`, which is validated outside the scope of this method; that's rather fragile.

ghokun replied:
Replaced type check with symbol check. Moved symbol capture outside.

Stephan202 commented on error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocation.java:
```diff
@@ -0,0 +1,185 @@
+    Type nameInvocationReceiverType = ASTHelpers.getReceiverType(enumValueArgument);
+    if (!ASTHelpers.isSameType(switchExpressionType, nameInvocationReceiverType, state)) {
+      return valuesOfReceiver;
+    }
```
IIUC, here we shouldn't check whether these guys are the same type, but in fact point to the same symbol. Let's also make sure to add a counter example test case for this.

ghokun replied:
Implemented symbol comparison.

ghokun replied:
Covered here: https://github.com/PicnicSupermarket/error-prone-support/pull/2031/files#r3266069897

Stephan202 reviewed (commented):
Quick partial review while @rickie and I were checking the code. Cool stuff @ghokun!

ghokun committed:
Replace type check with symbol check on switch statements

ghokun commented on error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocationTest.java:
```diff
@@ -0,0 +1,205 @@
+            "      }",
+            "    }",
+            "",
+            "    if (d instanceof D(B deconstructedB)) {",
```
There are also potential cases which I ignored on purpose such as:


```java
enum A {ONE};
enum B {ONE, TWO};
record D(B b) {}

D d = new D(B.ONE);
switch (d) {
  case D(B deconstructedB) when B == ONE -> ..
  case D(B deconstructedB) when B == TWO -> ..
};
```

I think this will complicate things even further. Should we place it inside the ignored part?

ghokun committed:
Remove unused argument

ghokun commented on error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocationTest.java:
```diff
@@ -0,0 +1,205 @@
+            "    Enum.valueOf(null, getClass().getSimpleName());",
+            "    List.of(\"\").stream().map(A::valueOf);",
+            "",
+            "    // Colon-style fall-through: `case FOUR` reaches the invocation, but only the",
```
Placed here to be explicit about the ignored case, if you see an easy way to implement it or that it is possible, let me know.

ghokun committed:
Cover case when symbols are different and receiver is limited by member select

ghokun committed:
Fix test

ghokun commented on error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/UnvalidatedEnumValueOfInvocationTest.java:
```diff
@@ -0,0 +1,220 @@
+            "      // BUG: Diagnostic matches: MISSING_VALUE",
+            "        consume(A.valueOf(B.FOUR.name()));",
+            "      // BUG: Diagnostic matches: MISSING_VALUE",
+            "        consume(Enum.valueOf(A.class, B.FIVE.name()));",
```
- Switch Cases are valid
- Types match
- Symbols do not match
- Value from the provided enum is a MemberSelectTree which is not present in A.class

ghokun committed:
Fix formatting of text wall

ghokun committed:
Remove extra curly bracket
